src/mlpack/core/optimizers/sdp/primal_dual_impl.hpp

```diff
@@ -0,0 +1,386 @@
+ *   SIAM J. Optim. 1998.
+ *   https://www.cs.nyu.edu/overton/papers/pdffiles/pdsdp.pdf
+ *
+ * We will refer to this paper as [AHO98] in this file.
+ *
+ * Note there are many optimizations that still need to be implemented. See the
+ * code comments for more details.
+ */
+#ifndef __MLPACK_CORE_OPTIMIZERS_SDP_PRIMAL_DUAL_IMPL_HPP
+#define __MLPACK_CORE_OPTIMIZERS_SDP_PRIMAL_DUAL_IMPL_HPP
+
+#include "primal_dual.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template <typename SDPType>
+PrimalDualSolver<SDPType>::PrimalDualSolver(const SDPType& sdp)
+  : sdp(sdp),
+    initialX(arma::eye<arma::mat>(sdp.N(), sdp.N())),
+    initialYsparse(arma::ones<arma::vec>(sdp.NumSparseConstraints())),
+    initialYdense(arma::ones<arma::vec>(sdp.NumDenseConstraints())),
+    initialZ(arma::eye<arma::mat>(sdp.N(), sdp.N())),
+    tau(0.99),
+    normXzTol(1e-7),
+    primalInfeasTol(1e-7),
+    dualInfeasTol(1e-7),
+    maxIterations(1000)
+{
+
+}
+
+template <typename SDPType>
+PrimalDualSolver<SDPType>::PrimalDualSolver(const SDPType& sdp,
+                                            const arma::mat& initialX,
+                                            const arma::vec& initialYsparse,
+                                            const arma::vec& initialYdense,
+                                            const arma::mat& initialZ)
+  : sdp(sdp),
+    initialX(initialX),
+    initialYsparse(initialYsparse),
+    initialYdense(initialYdense),
+    initialZ(initialZ),
+    tau(0.99),
+    normXzTol(1e-7),
+    primalInfeasTol(1e-7),
+    dualInfeasTol(1e-7),
+    maxIterations(1000)
+{
+  arma::mat tmp;
+
+  if (initialX.n_rows != sdp.N() || initialX.n_cols != sdp.N())
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialX needs to be square n x n matrix"
+      << std::endl;
+
+  if (!arma::chol(tmp, initialX))
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialX needs to be symmetric positive definite"
+      << std::endl;
+
+  if (initialYsparse.n_elem != sdp.NumSparseConstraints())
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialYsparse needs to have the same length as the number of sparse constraints"
+      << std::endl;
+
+  if (initialYdense.n_elem != sdp.NumDenseConstraints())
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialYdense needs to have the same length as the number of dense constraints"
+      << std::endl;
+
+  if (initialZ.n_rows != sdp.N() || initialZ.n_cols != sdp.N())
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialZ needs to be square n x n matrix"
+      << std::endl;
+
+  if (!arma::chol(tmp, initialZ))
+    Log::Fatal << "PrimalDualSolver::PrimalDualSolver(): "
+      << "initialZ needs to be symmetric positive definite"
+      << std::endl;
+}
+
+/**
+ * alphahat = sup{ alphahat : A + dA is psd }
+ */
+static inline double
+AlphaHat(const arma::mat& A, const arma::mat& dA)
```

Comment:
This is only used once; why not just inline it into Alpha()?  (which is pretty short anyway)